src/cpp/core/r_util/RProjectFile.cpp

```diff
@@ -311,6 +312,54 @@ std::ostream& operator << (std::ostream& stream, const YesNoAskValue& val)
    return stream ;
 }
 
+Error findProjectFile(FilePath filePath,
+                      FilePath* pProjPath)
+{
+   // check to see if we already have a .Rproj file
+   if (filePath.extension() == ".Rproj")
+   {
+      *pProjPath = filePath;
+      return Success();
+   }
+   
+   if (!filePath.isDirectory())
+      filePath = filePath.parent();
+   
+   // no .Rproj file found; scan parent directories
+   for (; filePath.exists(); filePath = filePath.parent())
```

Comment:
I think exists() logs an error on permissions (see https://github.com/rstudio/rstudio/blob/master/src/cpp/core/FilePath.cpp#L237-L253) so this has the potential to create log noise in shared projects or other situations wherein you're opening a file but don't necessarily have read access to all of the directories which contain it. One possible solution might be to only do this if you can find a common ancestor path (and to terminate the parent scan at that point).